### PR TITLE
Add native (DeepNVMe) host-memory pinning backend for accelerators

### PR DESCRIPTION
**Pull request by sfc-gh-truwase**
## Summary
Adds a native host-memory pinning backend, selectable via the `DS_PIN_MEMORY_BACKEND` environment variable (defaults to `torch`). When set to `native`, CPU memory is page-locked through the standalone DeepSpeed `pin_memory` op (`PinMemoryBuilder` / `pin_handle`, `posix_memalign` + `mlock`) instead of `torch.pin_memory()`.

Stacked on #8236 (standalone `pin_memory` op, now on `master`). Native allocations go through `pin_handle`, so DeepNVMe I/O handles recognize them via the process-wide manager and skip bounce buffers — without requiring libaio / AIO worker threads.

- **New `deepspeed/utils/pin_memory.py`**: a process-wide shared `NativePinnedMemory` manager that pins CPU memory, tracks pinned pointer ranges (so slices/views report as pinned), tags buffers with `.ds_pinned`, supports `make_copy`/`match_shape`, and frees on unpin. It fails early with a clear error if the `pin_memory` op cannot be built (no silent torch fallback). Native pins also use a `weakref` finalizer so GC releases mlocked pages when tensors are dropped without an explicit unpin.
- **Accelerator owns dispatch**: `pin_memory` drops `align_bytes` and gains `make_copy`/`match_shape`; `is_pinned` is FakeTensor/meta-tensor safe; new `unpin_memory` (native frees, torch no-op). Subclasses retain only the device-specific `_torch_pin_memory`/`_torch_is_pinned` primitives. Preserves master's `track_pinned_memory` accounting (CPU torch no-op still bypasses it).
- **Consolidation**: XPU's bespoke `align_bytes=0` path is folded into the shared native backend.
- **Callers**: `compile` paths route through `get_accelerator()`. Swap-tensor buffers continue to allocate via I/O handles; with the shared manager they interoperate with native-pinned tensors. ZeRO / ZenFlow `destroy()` explicitly unpins optimizer-owned CPU-offload buffers under the native backend.
- **Docs**: Host Memory Pinning section under RTD Memory Usage (`docs/code-docs/source/memory.rst`).
- **Tests**: unit tests for the native manager, accelerator pinning APIs, destroy-path unpin, and cross-op recognition with AIO.

## Test plan
- [x] Rebased onto `master` after #8236 merge; retargeted `NativePinnedMemory` from `AsyncIOBuilder` → `PinMemoryBuilder`.
- [x] `pre-commit` on changed files.
- [x] Focused UTs on GPU (`tunji-h200-n1g2-ds2-0`, job `20260809T183311Z`): `tests/unit/v1/pin_memory/` + `tests/unit/v1/accelerator/test_accelerator.py` + `tests/unit/v1/nvme/test_pinned_manager.py` — **30 passed**.
- [x] Bounce-buffer / cross-op smoke: under `DS_PIN_MEMORY_BACKEND=native`, a `pin_handle` buffer is `is_pinned` on a separate AIO handle.

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Comment by delock** [comment]
I think host memory pinning accouting is needed.  I made an attempt in the following PR (in order to justfy pin_memory as default offload option).
https://github.com/deepspeedai/DeepSpeed/pull/8207/changes

**Comment by sfc-gh-truwase** [comment]
Addressed Codex review ([r3713658448](https://github.com/deepspeedai/DeepSpeed/pull/8211#discussion_r3713658448)) by landing #8212 first, then rebasing this PR onto updated `master` (`f7934bc8`).

With the process-wide shared `deepspeed_pin_tensor_t` manager, buffers allocated via `NativePinnedMemory` (`new_cpu_locked_tensor`) are registered in the same manager every I/O handle consults. Verified on a GPU node: a native-pinned buffer (and a narrow of it) is reported `is_pinned` by a *different* AIO handle than the allocator, so DeepNVMe will skip the bounce-buffer path.

File: deepspeed/utils/pin_memory.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        try:
+            from deepspeed.ops.op_builder import AsyncIOBuilder
+            self._handle = AsyncIOBuilder().load().aio_handle(128 * 1024, 8, False, False, 1)
+        except Exception as e:
+            raise RuntimeError(
+                "DS_PIN_MEMORY_BACKEND=native requires the async-io (AIO) op, which failed to build/load.") from e
+
+    def pin(self, tensor, make_copy=True, match_shape=True):
+        numel = tensor.numel()
+        locked = self._handle.new_cpu_locked_tensor(numel, tensor)[:numel]
```
[the remark this answers is not in the document]
Comment (sfc-gh-truwase) [reply deepspeed/utils/pin_memory.py]: Fixed in 780f7864. `NativePinnedMemory.pin()` now attaches a `weakref.finalize` to the returned tensor that frees the page-locked allocation (`free_cpu_locked_tensor`) when the tensor is garbage-collected — matching `torch.pin_memory()`'s lifetime, so callers that never call `unpin_memory()` no longer accumulate mlocked host memory. An explicit `unpin()` cancels the finalizer (`.detach()`) to avoid a redundant free. Added `test_pin_freed_on_gc` (drops the tensor, forces GC, asserts the range is released); the full pin-memory suite is 22/22 on a GPU node.

**Comment by sfc-gh-truwase** [comment]
Addressed the native pin lifetime concern with two complementary paths:

1. **`weakref.finalize`** on `NativePinnedMemory.pin()` (already landed) so GC frees mlocked allocations when tensors are dropped without an explicit unpin.
2. **Explicit `accelerator.unpin_memory()` in ZeRO/`ZenFlow` `destroy()`** for optimizer-owned CPU-offload pins (this commit), so teardown does not wait on GC.

Coverage:
- ZeRO-1/2, ZeRO-3, and ZenFlow overrides unpin their owned offload buffers in `destroy()`.
- Param-offload partition pins that are rebound to flattened views are left to the finalizer (unpinning those views is ineffective / unsafe).
- New UTs in `tests/unit/v1/pin_memory/test_destroy_unpin.py` (5/5 passed on GPU): native ZeRO-2/3 free all optimizer pins; native ZeRO-3+param-offload decreases ranges; torch backend destroy remains a clean no-op.

File: deepspeed/utils/pin_memory.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        _shared_native_pins = NativePinnedMemory()
+    return _shared_native_pins
+
+
+def get_active_native_pinned_memory():
+    # Returns the shared manager when ``DS_PIN_MEMORY_BACKEND=native``; otherwise
+    # None so the accelerator uses its (torch) pinning path. When native is
+    # selected but the async-io ops cannot be built, constructing the manager
+    # raises rather than silently falling back.
+    if os.environ.get("DS_PIN_MEMORY_BACKEND", "torch").strip().lower() != "native":
```
[the remark this answers is not in the document]
Comment (sfc-gh-truwase) [reply deepspeed/utils/pin_memory.py:132]: Documented under RTD **Memory Usage** as [Host Memory Pinning](https://github.com/deepspeedai/DeepSpeed/blob/tjruwase/native-pin-memory/docs/code-docs/source/memory.rst) (`docs/code-docs/source/memory.rst`):

- `DS_PIN_MEMORY_BACKEND=torch|native` selection and defaults
- async-io build / memlock requirements for `native`
- accelerator `pin_memory` / `is_pinned` / `unpin_memory` APIs
- native lifetime (`weakref` finalizer + explicit unpin in ZeRO/`destroy()`)

Also refreshed the older **Pinned Memory** accounting notes to the current `offload_*.pin_memory` config knobs and cross-linked them to this section.

**Comment by sfc-gh-truwase** [comment]
@delock , @stas00, @tohtana  see this page for docs preview of this feature: 
 https://deepspeed.readthedocs.io/en/rtd-staging/memory.html#host-memory-pinning

**Comment by sfc-gh-truwase** [comment]
@delock Accounting from #8207 is now integrated into this PR (merged latest `master` in `0cca5393`).

`get_accelerator().pin_memory()` always goes through `track_pinned_memory` for both the default `torch` backend and `DS_PIN_MEMORY_BACKEND=native` (native tracks the returned locked buffer’s nbytes). The CPU accelerator still bypasses accounting on its torch no-op path (nothing is page-locked), and defers to the ABC — with tracking — when the native backend is active.

Pinned-memory volume notes in the RTD Memory Usage page also include the #8207 updates, alongside the Host Memory Pinning section for backend selection.

File: deepspeed/utils/pin_memory.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    # rather than in the accelerator, which only owns the device-specific torch
+    # pinning path.
+    def __init__(self):
+        # base address -> end address; keyed by base so unpin is an O(1) delete.
+        self._ranges = {}
+        # Fail early: native pinning is useless without the async-io handle, so
+        # surface the build/load failure here instead of silently degrading.
+        try:
+            from deepspeed.ops.op_builder import AsyncIOBuilder
+            self._handle = AsyncIOBuilder().load().aio_handle(128 * 1024, 8, False, False, 1)
```
[the remark this answers is not in the document]
Comment (sfc-gh-truwase) [reply deepspeed/utils/pin_memory.py]: Addressed by landing #8212 (shared process-wide `deepspeed_pin_tensor_t`) and rebasing this PR onto that `master`. `NativePinnedMemory` allocates via `new_cpu_locked_tensor`, so every AIO/GDS handle’s `is_pinned` / bounce-buffer check sees the same manager. Verified: a native-pinned buffer (and a narrow of it) is `is_pinned` on a different AIO handle than the allocator.

**Comment by sfc-gh-truwase** [comment]
Converted to draft. This PR will be rebased/stacked on top of #8236 (standalone `pin_memory` op split from async-io) once that lands or is ready as the base.

**Comment by sfc-gh-truwase** [comment]
Resumed after #8236 landed:

- Rebased `tjruwase/native-pin-memory` onto current `master` (includes standalone `pin_memory` op).
- Retargeted `NativePinnedMemory` to `PinMemoryBuilder().load().pin_handle()` (no longer depends on async-io / libaio).
- Merged Host Memory Pinning RTD with torch vs native backend table (build dep = **pin_memory** op).
- Preserved `track_pinned_memory` accounting from master.
- Verified: **30 passed** on GPU (pin_memory + accelerator + pinned_manager), including cross-op recognition.

Ready for review.

File: accelerator/xpu_accelerator.py [hunk 4]
```diff
@@ -229,26 +228,12 @@ def IntTensor(self):
     def LongTensor(self):
         return functools.partial(torch.tensor, dtype=torch.long, device=self._name)
 
-    def _pin_memory(self, tensor, align_bytes=1):
```
Comment (delock) [inline accelerator/xpu_accelerator.py:232]: Is this mean XPU accelerator needs to explicitly set `DS_PIN_MEMORY_BACKEND` to `native` with intended page-locked host buffer?

@rogerxfeng8 to confirm wheter XPU still relies on `new_cpu_locked_tensor` to work in certain scenario.

Comment (sfc-gh-truwase) [reply accelerator/xpu_accelerator.py:232]: @delock @rogerxfeng8 yes it will be great to confirm the best support for XPU.

File: deepspeed/utils/pin_memory.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        if match_shape:
+            locked = locked.view(tensor.shape)
+        begin = base.data_ptr()
+        self._ranges[begin] = begin + numel * tensor.element_size()
+        locked.ds_pinned = True
+        # Match torch.pin_memory lifetime semantics: free the page-locked
+        # allocation when the returned tensor is dropped, so call sites that never
+        # call unpin() do not accumulate mlocked host memory.
+        self._finalizers[begin] = weakref.finalize(locked, self._release, self._handle, base, self._ranges,
+                                                   self._finalizers, begin)
```
[the remark this answers is not in the document]
Comment (sfc-gh-truwase) [reply deepspeed/utils/pin_memory.py]: Fixed in 7df48c5b. The `weakref.finalize` is now tied to the allocation root (`base`), not the returned view. Every slice/view keeps `base` alive through `._base`, so the mlocked region is freed only after the returned tensor **and** all its aliases are gone — a live view can no longer alias a freed allocation. Freeing is done by address (see below) so `base` is not pinned alive by the finalizer itself. Added `test_allocation_survives_until_all_views_dropped`.

File: deepspeed/utils/pin_memory.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        if getattr(tensor, "ds_pinned", False):
+            return True
+        if not self._has_real_storage(tensor):
+            return False
+        ptr = tensor.data_ptr()
+        return any(begin <= ptr < end for begin, end in self._ranges.items())
+
+    def unpin(self, tensor):
+        # After freeing, using ``tensor`` is a use-after-free and must be avoided.
+        begin = tensor.data_ptr()
```
[the remark this answers is not in the document]
Comment (sfc-gh-truwase) [reply deepspeed/utils/pin_memory.py]: Fixed in 7df48c5b. `pin()` now records the owning allocation address on the returned tensor (`ds_pin_base`), and both `unpin()` and the finalizer free by that address via the new `pin_handle.free_cpu_locked_tensor_by_ptr` / `deepspeed_pin_tensor_t::free(void*)`. So even after Stage-3 `offload_states()` redirects `grad_partitions_flat_buffer.data`, `destroy()` releases the original locked range deterministically instead of leaving it until GC. Added `test_unpin_frees_original_after_data_redirect`.

File: deepspeed/utils/pin_memory.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
+# SPDX-License-Identifier: Apache-2.0
```
[the remark this answers is not in the document]
Comment (sfc-gh-truwase) [reply deepspeed/utils/pin_memory.py:1]: All commits on this branch carry a `Signed-off-by` trailer (created with `--signoff`); the DCO / required check is green. This looks like a false positive from reviewing the diff in isolation.

**Review by delock — APPROVED** [review]

File: accelerator/xpu_accelerator.py (hunk 4 above)
Comment (sfc-gh-truwase) [reply accelerator/xpu_accelerator.py:232]: @delock @rogerxfeng8 do we need changes for XPU?

Comment (delock) [reply accelerator/xpu_accelerator.py:232]: Hi  Tunji, I have validate this on arc770 with pytorch for xpu and met no UT failure.  I have approved the PR.  @rogerxfeng8 if you think any additional test is needed plz comment this week, thanks!
